test/Microsoft.AspNet.Server.Kestrel.FunctionalTests/PathBaseTests.cs

```diff
@@ -0,0 +1,97 @@
+        [InlineData("http://localhost:8791", "http://localhost:8791/something", "", "/something")]
+        [InlineData("http://localhost:8791/base", "http://localhost:8791", "", "/")]
+        [InlineData("http://localhost:8791/base", "http://localhost:8791/", "", "/")]
+        [InlineData("http://localhost:8791/base", "http://localhost:8791/something", "", "/something")]
+        [FrameworkSkipCondition(RuntimeFrameworks.Mono, SkipReason = "Test hangs after execution on Mono.")]
+        public Task DefaultPathBaseIsEmpty(string registerAddress, string requestAddress, string expectedPathBase, string expectedPath)
+        {
+            return TestPathBase(registerAddress, requestAddress, expectedPathBase, expectedPath);
+        }
+
+        [ConditionalTheory]
+        [InlineData("http://localhost:8791", "http://localhost:8791", "", "/")]
+        [InlineData("http://localhost:8791", "http://localhost:8791/", "", "/")]
+        [InlineData("http://localhost:8791/", "http://localhost:8791", "", "/")]
+        [InlineData("http://localhost:8791/", "http://localhost:8791/", "", "/")]
+        [InlineData("http://localhost:8791/base", "http://localhost:8791/base/", "/base", "/")]
+        [InlineData("http://localhost:8791/base/", "http://localhost:8791/base", "/base", "")]
+        [InlineData("http://localhost:8791/base/", "http://localhost:8791/base/", "/base", "/")]
+        [FrameworkSkipCondition(RuntimeFrameworks.Mono, SkipReason = "Test hangs after execution on Mono.")]
+        public Task PathBaseNeverEndsWithSlash(string registerAddress, string requestAddress, string expectedPathBase, string expectedPath)
+        {
+            return TestPathBase(registerAddress, requestAddress, expectedPathBase, expectedPath);
+        }
+
+        private async Task TestPathBase(string registerAddress, string requestAddress, string expectedPathBase, string expectedPath)
+        {
+            var config = new ConfigurationBuilder().AddInMemoryCollection(
+                new Dictionary<string, string> {
+                    { "server.urls", registerAddress }
+                }).Build();
+
+            var builder = new WebHostBuilder(config)
+                .UseServerFactory("Microsoft.AspNet.Server.Kestrel")
+                .UseStartup(app =>
+                {
+                    app.Run(async context =>
+                    {
+                        await context.Response.WriteAsync(JsonConvert.SerializeObject(new
+                        {
+                            PathBase = context.Request.PathBase.Value,
+                            Path = context.Request.Path.Value
+                        }));
+                    });
+                });
+
+            using (var app = builder.Build().Start())
+            {
+                using (var client = new HttpClient())
+                {
+                    var response = await client.GetAsync(requestAddress);
```

Comment:
HttpClient will always turn http://localhost to http://localhost/, so you can't test that senario here.